openpgp/keys.go

```diff
@@ -131,23 +131,35 @@ func (e *Entity) EncryptionKey(now time.Time) (Key, bool) {
 // SigningKey return the best candidate Key for signing a message with this
 // Entity.
 func (e *Entity) SigningKey(now time.Time) (Key, bool) {
+	return e.SigningKeyById(now, 0)
+}
+
+// SigningKeyById return the Key for signing a message with this
+// Entity and keyID.
+func (e *Entity) SigningKeyById(now time.Time, id uint64) (Key, bool) {
 	// Fail to find any signing key if the primary key has expired.
 	i := e.PrimaryIdentity()
 	primaryKeyExpired := e.PrimaryKey.KeyExpired(i.SelfSignature, now)
 	if primaryKeyExpired {
 		return Key{}, false
 	}
 
+	if e.PrivateKey != nil && e.PrivateKey.KeyId == id {
+		// If the id specified is the Entity.PrivateKey.KeyID lets skip looking at the subkeys
```

Comment:
Rather than short-circuiting here (which also skips checking whether the primary key is a valid signing key etc), could you add a check for the key ID to the last `if` in this function, and then if it doesn't match, we return `false` (at the end of this function). (If that causes an error, the app can then try again without passing `SigningKeyId`; that way it's less ambiguous that the option always forces that key to be used, or an error if that's not possible.)

Reply:
Makes sense